### ISSUE TITLE
Planning

### ISSUE THREAD

**Issue by MaxOstrowski** [potassco/clingo-server#1]
Just to do some idea collection of how the actual interface could look like and what the required features would be.

Features:

- multi-shot solving
- theory solving (clingo-dl, clingcon)
- multiple, (independent ?) clingo instances (easiest implementation uses one library include, but domain knowledge accumulates, server restarts would be required, fresh loading of clingo library would be helpful or another layer with 2 web servers, starting a fresh webserver internally for every clingo instance)

Interface requirements:

- create a (clingo) instance returning UID
- add theory N ("clingo-dl, "clingcon") to UID: check tefoli, how to do a rust interface for this
- add encoding E to UID
- add set of facts F to UID
- ground program N with parameter list P of UID
- set solve timeout of UID
- solve UID under assumption set of terms A
- get result (models) of UID
- add/(get) configuration of UID
- interrupt solve of UID
- (get statistics of UID)
- assign external term T to Bool B of UID
- release external term T of UID

I think this kind of interface covers a lot of use cases and still hides most of the clingo api.

Thoughts:

- not all functions can be used at any time (adding theory must be done before solving, result can only be obtained after solving etc...) maybe a getState could be used to check in which state the UID is. Usually a REST interface has to be stateless, but I think we can not achieve this behavior.
- how to communicate terms (input facts, model, etc..): pure strings would need to be parsed again, clingo symbols would require clingo library (which is bad)

Flaws:

- currently temporary files are created and used, this leads to undefined behavior if used at the same time from several instances

**Comment by sthiele** [comment]
I'm closing this in favor of more specific issues.
